Cache: GCC ASAN Build Warnings

# Steps to Reproduce the Problem

1. Download the file [Dockerfile.txt](https://github.com/ElektraInitiative/libelektra/files/3386130/Dockerfile.txt) and store it as `scripts/docker/debian/buster/Dockerfile`

2. Build the Docker image

   ```sh
   docker build \
          -t elektra-buster \
          --build-arg JENKINS_USERID=$(id -u) \
          --build-arg JENKINS_GROUPID=$(id -g) \
          -f scripts/docker/debian/buster/Dockerfile \
          scripts/docker/debian/buster/
   ```

3. Run the Docker image

   ```sh
   docker run \
          -it --rm --cap-add SYS_PTRACE \
          -v "$PWD:/home/jenkins/workspace" \
          -w /home/jenkins/workspace \
          elektra-buster
   ```

4. Create the build system

   ```sh
   mkdir build
   cd build
   cmake \
         -DBUILD_DOCUMENTATION=OFF \
         -DCOMMON_FLAGS=-Werror \
         -DENABLE_ASAN=ON \
         -DBINDINGS='ALL;-DEPRECATED' \
         -DPLUGINS='ALL;-DEPRECATED ' \
         -DTOOLS='ALL;-qt-gui' \
         -DENABLE_DEBUG=ON \
         -DKDB_DB_FILE=default.mmap \
         -DKDB_DB_INIT=elektra.mmap \
         -DKDB_DEFAULT_STORAGE=mmapstorage \
         ..
   ```

5. Compile Elektra

   ```sh
   cmake --build .
   ```

# Expected Result

The build system is able to translate Elektra without any warnings.

# Actual Result

GCC `8.3` reports a warning about a null destination pointer:

```
src/plugins/cache/cache.c: In function ‘elektraCacheSet’:
src/plugins/cache/cache.c:154:8: error: null destination pointer [-Werror=format-overflow=]
  len = sprintf (tmpFile, "%s", cacheFileName);
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

.

# System Information

- Elektra `master`
- Docker `2.0.0.3`

# Further Log Files and Output

- [Jenkins Build Job](https://build.libelektra.org/jenkins/blue/organizations/jenkins/libelektra/detail/PR-2819/11/pipeline/458)

--- a/doc/news/_preparation_next_release.md
+++ b/doc/news/_preparation_next_release.md
@@ -86,6 +86,7 @@ The following section lists news about the [modules](https://www.libelektra.org/
 - Add check of resolved filenames, fixes false cache hits. _(Mihael Pranjić)_
 - Skip all plugins and global plugins when we have a cache hit. _(Mihael Pranjić)_
 - Fix two data loss bugs when using `cache` with `multifile` resolver. _(Mihael Pranjić)_
+- Fix GCC ASAN build warnings #2820. _(Mihael Pranjić)_
 
 ### crypto / fcrypt
 

--- a/src/plugins/cache/cache.c
+++ b/src/plugins/cache/cache.c
@@ -151,7 +151,7 @@ static char * elektraGenTempFilename (char * cacheFileName)
 
 	tmpFilenameSize = strlen (cacheFileName) + POSTFIX_SIZE;
 	tmpFile = elektraCalloc (tmpFilenameSize);
-	len = sprintf (tmpFile, "%s", cacheFileName);
+	len = snprintf (tmpFile, strlen (cacheFileName), "%s", cacheFileName);
 
 	struct timeval tv;
 	memset (&tv, 0, sizeof (struct timeval));

--- a/src/plugins/cache/cache.c
+++ b/src/plugins/cache/cache.c
@@ -149,9 +149,10 @@ static char * elektraGenTempFilename (char * cacheFileName)
 	size_t len = 0;
 	size_t tmpFilenameSize = 0;
 
-	tmpFilenameSize = strlen (cacheFileName) + POSTFIX_SIZE;
+	size_t cacheFileNameSize = strlen (cacheFileName);
+	tmpFilenameSize = cacheFileNameSize + POSTFIX_SIZE;
 	tmpFile = elektraCalloc (tmpFilenameSize);
-	len = snprintf (tmpFile, strlen (cacheFileName), "%s", cacheFileName);
+	len = snprintf (tmpFile, cacheFileNameSize + 1, "%s", cacheFileName);
 
 	struct timeval tv;
 	memset (&tv, 0, sizeof (struct timeval));

